Add auto apply for theme

norgeindian opened the pull request:
Added automatic apply to theme templates

convenient commented on src/Ampersand/PatchHelper/Patchfile/Entry.php:
```diff
@@ -263,4 +263,19 @@ public function __toString()
+            $adaptedLines [] = str_replace('vendor_orig', 'vendor', $adaptedLine);
+        }
+        file_put_contents($tmpPatchFilePath, implode(PHP_EOL, $adaptedLines));
+        shell_exec('patch < ' . $tmpPatchFilePath . ' -p0 -F' . $fuzzFactor . ' --no-backup-if-mismatch -d'
```
So far the only other usage of `shell_exec` is executing `find`.

I've got a sanity check that `find` is installed at this location
https://github.com/AmpersandHQ/ampersand-magento2-upgrade-patch-helper/blob/master/src/functions.php#L3-L5

Would be good to have a similar sanity check for `patch`

norgeindian replied:
@convenient , thanks for your feedback.
Added a check

convenient commented:
This looks like an amazing PR. thanks for the work. I know you'll be slugging away at a lot of code, what do you think the odds are that you'd be able to cover this with a test of some variety?

convenient commented:
@hostep I believe I recall you were saying you were interesting in this kind of thing. Was wondering if you had any thoughts on this?

hostep commented:
This feature looks pretty awesome because it could potentially save a lot of manual work.
But I'd rather not see it being run automatically, there should be a new flag for this I believe?

However, I already tried to apply patches like this on a custom theme, and it only worked for about 5 to 10% of the overrides if I remember correctly. Because often it can't apply the patch due to some small change (like a `class` change in a phtml file or an extra attribute on a html element, or a translation change, or some small formatting changes, or because some html was changed in sort order and brought more to the bottom or to the top of a file, or because a part of the html was removed because it was unnecessary, or ...).

What could help, was trying to apply different hunks of a patch separately. Sometimes a patch consists of multiple hunks and one hunk can sometimes be applied automatically and another not, so maybe that's an option here, to try to get the hunks split up into individual patches and trying to apply those one by one?

We would then also need some sort of report of what it was able to apply and what not because the things it couldn't apply need then to be done manually.

I've also seen occurrences where part of the original phtml file was copied a second time in the same file and the change of a new version needed to be applied twice instead of once.
I've also seen a phtml file being copied entirely and given a new filename, some small changes were added and it was inserted using xml again. So that would also mean that both files needed to be updated with changes form a new Magento version.
But these kind of things probably can't be checked/patched automatically somehow, it would become very complex to try to implement such a thing.

--

Overall I'm all for such a feature, but it might be very difficult to get this right, but this might be a good first step! 🙂

convenient commented:
Thanks for your input @hostep, valuable as always.

>But I'd rather not see it being run automatically, there should be a new flag for this I believe?

A new flag sounds good I think. Agreed on that.

>To try to get the hunks split up into individual patches and trying to apply those one by one?

In this PR we're going over the patchfile lines and generating a patchfile and attempting to apply it. 
https://github.com/AmpersandHQ/ampersand-magento2-upgrade-patch-helper/blob/a9a32099ffe56ea75797b97e3f296fd40eabb4ad/src/Ampersand/PatchHelper/Patchfile/Entry.php#L271-L276

We already have code to get the hunks https://github.com/AmpersandHQ/ampersand-magento2-upgrade-patch-helper/blob/5d8fc8e83153fd45562105b351d2e303dea42049/src/Ampersand/PatchHelper/Patchfile/Entry.php#L54-L81

Perhaps instead of iterating over the lines and generating one patchfile to apply, we could try generating multiple patchfiles and trying to apply them? I'm not sure how that would work with the context and line numbers though as you go through the hunks the line numbers will start to shift further from what the unified diff format expects? Would possibly need a large fuzz.

> I've also seen a phtml file being copied entirely and given a new filename, some small changes were added and it was inserted using xml again .... But these kind of things probably can't be checked/patched automatically somehow, it would become very complex to try to implement such a thing.

Yes I can't see how we could easily do that without introducing some static analysis of the code base, looking for similar line contents etc. Seems a lot of work.

norgeindian commented:
I changed the type to be optional now. So everyone can decide, if he wants to use the auto update functionality or not.
In my local tests this approach already worked for a lot of templates.
In my eyes it is pretty unlikely that the local rewrite of a template adapted exactly the part, which is now updated.
So in most of the cases this should work.
Of course it does not help anymore, when you changed most of the template file, but in this case I don't see a chance to automate that anyway.

@convenient, @hostep , please be so kind and take a look at this again and please let me know what you still miss to merge my pull request. Thanks.

convenient commented:
@norgeindian Thanks for the update. To me this looks like a good first pass of this functionality. Can you address the code style issues which are causing the failure please?

convenient commented:
Hello @norgeindian 

I have pushed some commits. You can see in b7904f5d052111fd305d81e050e92c67e3c25fe9 I had to refactor your code a bit in order to get the patch to apply, I could not get it to accept the input of my test otherwise.

Can you please review this? I have the test going green and the input patch file looks to be of the expected format, but I never got your original code working so I am not sure if I'm missing anything.

# Test case

We have an override of a template file like this (only change is the comment `<!-- We have overridden this file -->
`)

https://github.com/AmpersandHQ/ampersand-magento2-upgrade-patch-helper/blob/b7904f5d052111fd305d81e050e92c67e3c25fe9/dev/TestModule/app/design/frontend/Ampersand/theme/Magento_Bundle/templates/js/components.phtml#L1-L9

Given the input patchfile 
https://github.com/AmpersandHQ/ampersand-magento2-upgrade-patch-helper/blob/b7904f5d052111fd305d81e050e92c67e3c25fe9/dev/phpunit/functional/resources/template-change.diff#L1-L9

The code generates this new `tmp.patch`

```
diff -ur app/design/frontend/Ampersand/theme/Magento_Bundle/templates/js/components.phtml app/design/frontend/Ampersand/theme/Magento_Bundle/templates/js/components.phtml
--- app/design/frontend/Ampersand/theme/Magento_Bundle/templates/js/components.phtml     2020-04-13 17:35:38.000000000 +0100
+++ app/design/frontend/Ampersand/theme/Magento_Bundle/templates/js/components.phtml          2020-07-31 12:30:43.000000000 +0100
@@ -4,5 +4,6 @@
  * See COPYING.txt for license details.
  */
 ?>
+<!-- Upgrade file changes get patched here!-->
 <?= $block->getChildHtml() ?>
```

Which applies successfully with fuzzing to match this file https://github.com/AmpersandHQ/ampersand-magento2-upgrade-patch-helper/blob/b7904f5d052111fd305d81e050e92c67e3c25fe9/dev/phpunit/functional/expected_output/auto-apply-patch.txt#L1-L9

To me this seems it is working?

norgeindian commented:
@convenient , just tested your code. Worked as expected and thanks a lot for adding a test.
I would really appreciate if you could merge the pull request.

convenient commented:
merged and tagged @norgeindian thank you for your contributions
